File: archiver/src/main/scala/farjs/archiver/AddToArchPopup.scala
package farjs.archiver

import farjs.ui._
import farjs.ui.border._
import farjs.ui.popup.ModalContent._
import farjs.ui.popup._
import farjs.ui.theme.Theme
import scommons.react._
import scommons.react.hooks._

import scala.scalajs.js

case class AddToArchPopupProps(zipName: String,
                               action: AddToArchAction,
                               onAction: String => Unit,
                               onCancel: () => Unit)

object AddToArchPopup extends FunctionComponent[AddToArchPopupProps] {

  private[archiver] var modalComp: UiComponent[ModalProps] = Modal
  private[archiver] var textLineComp: ReactClass = TextLine
  private[archiver] var textBoxComp: UiComponent[TextBoxProps] = TextBox
  private[archiver] var horizontalLineComp: ReactClass = HorizontalLine
  private[archiver] var buttonsPanelComp: ReactClass = ButtonsPanel

  protected def render(compProps: Props): ReactElement = {
    val props = compProps.wrapped
    val (zipName, setZipName) = useState(props.zipName)
    val (width, height) = (75, 8)
    val contentWidth = width - (paddingHorizontal + 2) * 2
    val contentLeft = 2
    val theme = Theme.useTheme().popup.regular

    val onAction: js.Function0[Unit] = { () =>
      if (zipName.nonEmpty) {
        props.onAction(zipName)
      }
    }
    
    val actions = js.Array(
      ButtonsPanelAction(s"[ ${props.action} ]", onAction),
      ButtonsPanelAction("[ Cancel ]", props.onCancel)
    )

    <(modalComp())(^.plain := ModalProps(s"${props.action} files to archive", width, height, theme, props.onCancel))(
      <(textLineComp)(^.plain := TextLineProps(
        align = TextAlign.left,
        left = contentLeft,
        top = 1,
        width = contentWidth,
        text = s"${props.action} to zip archive:",
        style = theme,
        padding = 0
      ))(),
      <(textBoxComp())(^.plain := TextBoxProps(
        left = contentLeft,
        top = 2,
        width = contentWidth,
        value = zipName,
        onChange = { value =>
          setZipName(value)
        },
        onEnter = onAction
      ))(),
      
      <(horizontalLineComp)(^.plain := HorizontalLineProps(
        left = 0,
        top = 3,
        length = width - paddingHorizontal * 2,
        lineCh = SingleChars.horizontal,
        style = theme,
        startCh = DoubleChars.leftSingle,
        endCh = DoubleChars.rightSingle
      ))(),
      <(buttonsPanelComp)(^.plain := ButtonsPanelProps(
        top = 4,
        actions = actions,
        style = theme,
        margin = 2
      ))()
    )
  }
}

File: archiver/AddToArchPopup.mjs
/**
 * @import { ButtonsPanelAction } from "@farjs/ui/ButtonsPanel.mjs"
 */
import React, { useState } from "react";
import Theme from "@farjs/ui/theme/Theme.mjs";
import TextAlign from "@farjs/ui/TextAlign.mjs";
import TextLine from "@farjs/ui/TextLine.mjs";
import TextBox from "@farjs/ui/TextBox.mjs";
import ButtonsPanel from "@farjs/ui/ButtonsPanel.mjs";
import HorizontalLine from "@farjs/ui/border/HorizontalLine.mjs";
import DoubleChars from "@farjs/ui/border/DoubleChars.mjs";
import SingleChars from "@farjs/ui/border/SingleChars.mjs";
import ModalContent from "@farjs/ui/popup/ModalContent.mjs";
import Modal from "@farjs/ui/popup/Modal.mjs";

const h = React.createElement;

/**
 * @typedef {"Add" | "Copy" | "Move"} AddToArchAction
 */

/**
 * @typedef {{
 *  readonly zipName: string;
 *  readonly action: AddToArchAction;
 *  onAction(zipName: string): void;
 *  onCancel(): void;
 * }} AddToArchPopupProps
 */

/**
 * @param {AddToArchPopupProps} props
 */
const AddToArchPopup = (props) => {
  const {
    modalComp,
    textLineComp,
    textBoxComp,
    horizontalLineComp,
    buttonsPanelComp,
  } = AddToArchPopup;

  const [zipName, setZipName] = useState(props.zipName);

  const [width, height] = [75, 8];
  const contentWidth = width - (ModalContent.paddingHorizontal + 2) * 2;
  const contentLeft = 2;
  const theme = Theme.useTheme().popup.regular;

  const onAction = () => {
    if (zipName.length > 0) {
      props.onAction(zipName);
    }
  };

  /** @type {readonly ButtonsPanelAction[]} */
  const actions = [
    { label: `[ ${props.action} ]`, onAction: onAction },
    { label: "[ Cancel ]", onAction: props.onCancel },
  ];

  return h(
    modalComp,
    {
      title: `${props.action} files to archive`,
      width,
      height,
      style: theme,
      onCancel: props.onCancel,
    },
    h(textLineComp, {
      align: TextAlign.left,
      left: contentLeft,
      top: 1,
      width: contentWidth,
      text: `${props.action} to zip archive:`,
      style: theme,
      padding: 0,
    }),
    h(textBoxComp, {
      left: contentLeft,
      top: 2,
      width: contentWidth,
      value: zipName,
      onChange: setZipName,
      onEnter: onAction,
    }),

    h(horizontalLineComp, {
      left: 0,
      top: 3,
      length: width - ModalContent.paddingHorizontal * 2,
      lineCh: SingleChars.horizontal,
      style: theme,
      startCh: DoubleChars.leftSingle,
      endCh: DoubleChars.rightSingle,
    }),
    h(buttonsPanelComp, {
      top: 4,
      actions,
      style: theme,
      margin: 2,
    })
  );
};

AddToArchPopup.displayName = "AddToArchPopup";
AddToArchPopup.modalComp = Modal;
AddToArchPopup.textLineComp = TextLine;
AddToArchPopup.textBoxComp = TextBox;
AddToArchPopup.horizontalLineComp = HorizontalLine;
AddToArchPopup.buttonsPanelComp = ButtonsPanel;

export default AddToArchPopup;
